engine/access/subscription/blocks_watcher.go

```diff
@@ -0,0 +1,209 @@
+		if err != nil {
+			return 0, rpc.ConvertStorageError(fmt.Errorf("could not get header for height %d: %w", startHeight, err))
+		}
+
+		if blockStatus == flow.BlockStatusFinalized {
+			return header.Height, nil
+		}
+	}
+
+	lastSealed, err := h.state.Sealed().Head()
+	if err != nil {
+		return 0, status.Errorf(codes.Internal, "could not get latest sealed block: %v", err)
+	}
+
+	// this checking if start block is sealed
+	if header != nil {
+		if header.Height > lastSealed.Height {
+			return 0, status.Errorf(codes.InvalidArgument, "provided start block must be sealed. Latest sealed block %v with the height %d", lastSealed.ID(), lastSealed.Height)
+		}
+
+		return header.Height, nil
+	}
+
+	// if no start block was provided, use the latest sealed block
+	return lastSealed.Height, nil
+}
+
+// GetHighestHeight returns the highest height based on the specified block status. Only flow.BlockStatusFinalized and flow.BlockStatusSealed allowed.
+func (h *BlocksWatcher) GetHighestHeight(blockStatus flow.BlockStatus) (uint64, error) {
+	switch blockStatus {
+	case flow.BlockStatusFinalized:
+		return h.finalizedHighestHeight.Value(), nil
+	case flow.BlockStatusSealed:
+		return h.sealedHighestHeight.Value(), nil
+	case flow.BlockStatusUnknown:
+		return 0, status.Errorf(codes.InvalidArgument, "could not get highest height for block with unknown status")
+	}
+
+	return 0, status.Errorf(codes.InvalidArgument, "could not get highest height for invalid status")
+}
+
+// SetFinalizedHighestHeight sets the highest finalized block height.
+func (h *BlocksWatcher) SetFinalizedHighestHeight(height uint64) bool {
+	return h.finalizedHighestHeight.Set(height)
+}
+
+// SetSealedHighestHeight sets the highest sealed block height.
+func (h *BlocksWatcher) SetSealedHighestHeight(height uint64) bool {
+	return h.sealedHighestHeight.Set(height)
+}
+
+// ProcessSubscriptionOnFinalizedBlock processes the subscription logic when a block is finalized.
+func (h *BlocksWatcher) ProcessSubscriptionOnFinalizedBlock(finalizedHeader *flow.Header) error {
+	if ok := h.SetFinalizedHighestHeight(finalizedHeader.Height); !ok {
+		h.log.Debug().Msg("finalized block already received")
+	}
+
+	// retrieve latest _finalized_ seal in the fork with head finalizedBlock and update last
+	// sealed height; we do _not_ bail, because we want to re-request approvals
+	// especially, when sealing is stuck, i.e. last sealed height does not increase
+	finalizedSeal, err := h.seals.HighestInFork(finalizedHeader.ID())
+	if err != nil {
+		return fmt.Errorf("could not retrieve finalizedSeal for finalized block %s", finalizedHeader.ID())
+	}
+	lastBlockWithFinalizedSeal, err := h.headers.ByBlockID(finalizedSeal.BlockID)
+	if err != nil {
+		return fmt.Errorf("could not retrieve last sealed block %v: %w", finalizedSeal.BlockID, err)
+	}
+
+	if ok := h.SetSealedHighestHeight(lastBlockWithFinalizedSeal.Height); !ok {
+		h.log.Debug().Msg("sealed block already received")
```

Comment:
again, it's unlikely that we will hit this but it's safe just to return here since it means that other thread has made more progress